Source/Core/Common/HandShake.h

```diff
@@ -0,0 +1,86 @@
+// Copyright 2018 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <array>
+#include <atomic>
+#include <mutex>
+#include <optional>
+
+template <typename Inner>
+class HandShake
+{
+public:
+  class ReaderGuard
+  {
+  public:
+    ReaderGuard(HandShake& par)
+        : parent(par), guard(par.sides[par.select.load(std::memory_order_acquire)].mutex)
+    {
+    }
+    // announces a switch
+    ~ReaderGuard()
+    {
+      parent.select.store(parent.select.load(std::memory_order_acquire) ^ 1,
+                          std::memory_order_release);
+    }
+    Inner& GetRef() { return parent.sides[parent.select.load(std::memory_order_relaxed)].inner; }
+
+  private:
+    HandShake& parent;
+    std::unique_lock<std::mutex> guard;
+  };
+  // dropping a value of this type will release the other side to the reader
+  class YieldGuard
+  {
+    friend class HandShake<Inner>;
+
+  public:
+    YieldGuard(HandShake& par) : parent(par), guard(parent.sides[parent.side].mutex) {}
+    /// returns the side that will be freed after dropping this guard
+    Inner& GetRef() { return parent.sides[1 ^ parent.side].inner; }
+
+  private:
+    HandShake& parent;
+    std::unique_lock<std::mutex> guard;
+  };
+
+  // reader interface
+  std::optional<ReaderGuard> TryRead();
+  ReaderGuard Wait() { return ReaderGuard(*this); }
+
+  // writer interface
+  Inner& GetWriter() { return sides[side].inner; }
+  std::optional<YieldGuard> Yield()
+  {
+    if (select.load(std::memory_order_acquire) != side)
+    {  // reader is still on the other side
+      return {};
+    }
+    side ^= 1;
+    auto res = std::make_optional<YieldGuard>(*this);
+    writerGuard.swap(res->guard);
```

Comment:
I am finding the logic behind this hard to understand. 

You create a `lock_guard` in `YieldGuard`, then hand it off to `HandShake`, then never release it until `HandShake` is destructed.

This doesn't seem to have a purpose.

Reply:
I don't see how that would deadlock, since the method returns early if select hasn't changed.

Reply:
That was my error, I apologise.

Sequence of events:

1. On the first call to `Yield`:
   1. `sides[0].mutex` is acquired by `lock_guard` in `YieldGuard`'s constructor on line 20.
   1. the guard is swapped with `writerGuard`
   1. `Yield` returns. We note that we have currently holding an active lock on `sides[0].mutex`.
2. Subsequent calls to `Yield` return `std::nullopt`.
3. A swap happens.
1. On the next call to `Yield`:
   1. `sides[1].mutex` is acquired by `lock_guard` in `YieldGuard`'s constructor.
   1. the guard is swapped with `writerGuard`
   1. `Yield` returns. We note that we have currently holding an active lock on `sides[0].mutex` and `sides[1].mutex`.
1. `sides[0].mutex` is eventually released when `YieldGuard` is destructed.
1. Repeat from 2 ad nauseum.